src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -806,15 +806,15 @@
       {:kind :invalid-subject-name
        :msg  (i18n/tru "Certificate names must be lower case.")}))
 
-  (when-not (re-matches #"\A[ -.0-~]+\Z" subject)
+  (when (.contains subject "*")
     (sling/throw+
       {:kind :invalid-subject-name
-       :msg  (i18n/tru "Subject contains unprintable or non-ASCII characters")}))
-
-  (when (.contains subject "*")
+       :msg  (i18n/tru "Subject contains a wildcard, which is not allowed: {0}" subject)}))
+  
+  (when-not (re-matches #"^([a-z0-9](?:(?:[a-z0-9-_]*|(?<!-)\.(?![-.]))*[a-z0-9]+)?)$" subject)
```

Comment:
I believe the dashes in the ranges that are supposed to be dashes need to be escaped:
`^([a-z0-9](?:(?:[a-z0-9\-_]*|(?<!-)\.(?![\-.]))*[a-z0-9]+)?)$`

Can you add some test cases to make sure those are accepted correctly?   I know we have a test around capital letters. I think more testing would be a great idea.